vote_persistence_splitbrain_test.go

```diff
@@ -0,0 +1,237 @@
+	// lastVoteTerm != currentTerm
+	require.NoError(t, victimStore.Set(keyLastVoteCand, []byte("old-candidate")))
+
+	// Capture stores and transport so we can restart over the same durable
+	// state.
+	oldStore := c.stores[victimIdx]
+	oldStable := c.stores[victimIdx]
+	oldSnap := c.snaps[victimIdx]
+	oldTrans := c.trans[victimIdx]
+	victimID := victim.localID
+
+	shutdownFuture := victim.Shutdown()
+	require.NoError(t, shutdownFuture.Error())
+
+	_ = oldTrans.Close()
+	addr := oldTrans.LocalAddr()
+	newAddr, newTrans := NewInmemTransport(addr)
+	require.Equal(t, addr, newAddr)
+
+	restartConf := *conf
+	restartConf.LocalID = victimID
+	restartConf.Logger = newTestLoggerWithPrefix(t, string(victimID)+"-restart")
+
+	restarted, err := NewRaft(&restartConf, &MockFSM{}, oldStore, oldStable, oldSnap, newTrans)
+	require.NoError(t, err)
+
+	// Update cluster bookkeeping to point at the restarted Raft instance.
+	c.rafts[victimIdx] = restarted
+	c.trans[victimIdx] = newTrans
+
+	// Verify the simulated crash state: currentTerm is up-to-date but
+	// lastVoteTerm is stale (represents crash between setCurrentTerm and persistVote)
+	require.Equal(t, curTermBefore, restarted.getCurrentTerm(), "currentTerm should survive crash")
+	lastVoteTermAfter, err := victimStore.GetUint64(keyLastVoteTerm)
+	require.NoError(t, err)
+	require.Less(t, lastVoteTermAfter, curTermBefore,
+		"simulated crash should leave lastVoteTerm < currentTerm (crash between setCurrentTerm and persistVote)")
+	t.Logf("Simulated crash state: currentTerm=%d, lastVoteTerm=%d (stale)", curTermBefore, lastVoteTermAfter)
+
+	// ---------------------------------------------------------------------
+	// Phase 3: Isolate the previous new leader and trigger a second
+	// RequestVote for the same term from another candidate (N1) to the
+	// restarted victim. The victim must NOT grant a second vote in that
+	// term even though its durable vote was lost.
+	// ---------------------------------------------------------------------
+
+	// Reconnect everyone first so that future partitions are well-defined.
+	c.FullyConnect()
+
+	// Now isolate the former new leader by putting it on the far side of the
+	// partition. This leaves {N1, restarted victim} able to communicate.
+	c.Partition([]ServerAddress{newLeader.localAddr})
+
+	// For safety, clear the victim's notion of any leader so the RequestVote
+	// is not rejected just because it believes some other leader exists.
+	restarted.setLeader("", "")
+
+	// Construct a RequestVote from N1 targeting the restarted victim.
+	// We use the victim's currentTerm to keep the vote in the same term.
+	victimLastIdx, victimLastTerm := restarted.getLastEntry()
+
+	candidateTrans := c.trans[idx1]
+	reqVote2 := RequestVoteRequest{
+		RPCHeader:          n1.getRPCHeader(),
+		Term:               restarted.getCurrentTerm(),
+		LastLogIndex:       victimLastIdx,
+		LastLogTerm:        victimLastTerm,
+		LeadershipTransfer: false,
+	}
```

Comment:
I haven't had a chance to go through in detail yet but this jumps out at me: doesn't the scenario assume that N1 is on the new term? Why would we get the index and term for N1's vote request from the victim, which we know is wrong?

Reply:
Actually, it's even worse than this because N1's `RequestVoteRequest.Term` should be for the _next_ term, with `LastLogTerm` being its current term.

```go
	// Construct a RequestVote from N1 targeting the restarted victim.
	n1LastIdx, n1LastTerm := n1.getLastEntry()

	candidateTrans := c.trans[idx1]
	reqVote2 := RequestVoteRequest{
		RPCHeader:          n1.getRPCHeader(),
		Term:               n1LastTerm + 1,
		LastLogIndex:       n1LastIdx,
		LastLogTerm:        n1LastTerm,
		LeadershipTransfer: false,
	}
```